source/features/useful-not-found-page.tsx

```diff
@@ -107,17 +112,50 @@ function init(): false | void {
 	for (let i = bar.children.length - 2; i >= 0; i--) {
 		void checkAnchor(bar.children[i] as HTMLAnchorElement);
 	}
+}
 
-	if (['tree', 'blob'].includes(parts[2])) {
-		// Object might be 410 Gone
-		void addCommitHistoryLink(bar);
+async function showHelpfulLinks(): Promise<void> {
+	const urlToFileOnDefaultBranch = await getUrlToFileOnDefaultBranch();
+	if (urlToFileOnDefaultBranch) {
+		select('h2')!.after(
+			<p className="container mt-4 text-center">
+				<a href={urlToFileOnDefaultBranch}>This {getType()}</a> exists on the default branch.
+			</p>,
+		);
+		return;
 	}
 
-	if (['tree', 'blob', 'edit'].includes(parts[2])) {
-		// File might not be available on the current branch
-		// GitHub already redirects /tree/ and /blob/ natively
-		void addDefaultBranchLink(bar);
+	const change = await getLatestChangeToFile();
+	if (!change) return;
+
+	const {fileInfo, commitInfo} = change;
+	const url = new GitHubURL(location.href);
+
+	url.assign({route: 'commits'});
+	const commitHistory = <a href={url.toString()}>Commit history</a>;
+	url.assign({route: 'blob', branch: commitInfo.parents[0].sha, filePath: url.filePath});
+	const lastVersionUrl = fileInfo.status === 'removed' ? fileInfo.blob_url : url.toString();
```

Comment:
I haven't tested this, but isn't `blob_url` also a 404 here?

Reply:
`blob_url` will be a link to the new file in case the file was renamed, but it will be the link to the last known version of the file if it has been deleted. In any case, I don't think it will be a 404.